sota-client/src/images.rs

```diff
@@ -108,39 +112,88 @@ pub struct ImageWriter {
 }
 
 impl ImageWriter {
-    /// Prepare to read an image file in chunks.
-    pub fn new(meta: ImageMeta, image_dir: String) -> Result<Self, Error> {
-        fs::create_dir_all(&image_dir)?;
-        let file = File::create(&format!("{}/{}", image_dir, meta.image_name))?;
-        file.set_len(meta.image_size)?;
-
-        Ok(ImageWriter {
+    /// Prepare to write an image file in chunks.
+    pub fn new(meta: ImageMeta, image_dir: String) -> Self {
+        ImageWriter {
             meta: meta,
             image_dir: image_dir,
             last_written: Utc::now(),
             chunks_written: HashSet::new(),
-        })
+        }
     }
 
-    /// Write a chunk of the image to the disk.
-    pub fn write_chunk(&mut self, data: &[u8], index: u64) -> Result<(), Error> {
-        trace!("writing chunk {} for {}", index, self.meta.image_name);
-        let path = format!("{}/{}", self.image_dir, self.meta.image_name);
-        let mut file = OpenOptions::new().write(true).open(path)?;
+    /// Write a single chunk directly to the output image.
+    pub fn write_direct(&mut self, data: &[u8], index: u64) -> Result<(), Error> {
```

Comment:
I'm not sure what happens here if you get the chunks out of order. Would you get some sort of "index out of bounds" error? Or is the file allocated to the full size when it's initialized?